lib/api/CAnomalyJobConfig.cc

```diff
@@ -520,6 +650,292 @@ void CAnomalyJobConfig::CAnalysisConfig::CDetectorConfig::parse(
+                                  : model::function_t::E_IndividualMinVelocity;
+        fieldRequired = true;
+    } else if (m_FunctionName == FUNCTION_MEAN_VELOCITY) {
+        m_Function = isPopulation ? model::function_t::E_PopulationMeanVelocity
+                                  : model::function_t::E_IndividualMeanVelocity;
+        fieldRequired = true;
+    } else if (m_FunctionName == FUNCTION_SUM_VELOCITY) {
+        m_Function = isPopulation ? model::function_t::E_PopulationSumVelocity
+                                  : model::function_t::E_IndividualSumVelocity;
+        fieldRequired = true;
+    } else {
+        LOG_ERROR(<< "Invalid function " << m_FunctionName << " specified");
+        return false;
+    }
+
+    // Validate
+    if (model::function_t::isPopulation(m_Function) && isPopulation == false) {
+        LOG_ERROR(<< "Function " << m_FunctionName << " requires an 'over' field");
+        return false;
+    }
+
+    if (isPopulation && model::function_t::isPopulation(m_Function) == false) {
+        LOG_ERROR(<< "Function " << m_FunctionName << " cannot be used with an 'over' field");
+        return false;
+    }
+
+    if (byFieldRequired && hasByField == false) {
+        LOG_ERROR(<< "Function " << m_FunctionName << " requires a 'by' field");
+        return false;
+    }
+
+    if (byFieldInvalid && hasByField) {
+        LOG_ERROR(<< "Function " << m_FunctionName << " cannot be used with a 'by' field");
+        return false;
+    }
+
+    if (fieldRequired && m_FieldName.empty()) {
+        LOG_ERROR(<< "Function " << m_FunctionName << " requires a field");
+        return false;
+    }
+
+    if (fieldInvalid && m_FieldName.empty() == false) {
+        LOG_ERROR(<< "Function " << m_FunctionName << " does not work on a field");
+        return false;
+    }
+
+    return true;
+}
+
+model_t::EExcludeFrequent
+CAnomalyJobConfig::CAnalysisConfig::CDetectorConfig::excludeFrequent() const {
+    if (m_OverHasExcludeFrequent) {
+        if (m_ByHasExcludeFrequent) {
+            return model_t::E_XF_Both;
+        } else {
+            return model_t::E_XF_Over;
+        }
+    } else {
+        if (m_ByHasExcludeFrequent) {
+            return model_t::E_XF_By;
+        }
+    }
+    return model_t::E_XF_None;
+}
+
+bool CAnomalyJobConfig::CAnalysisConfig::CDetectorConfig::decipherExcludeFrequentSetting() {
+
+    bool hasByField{m_ByFieldName.empty() == false};
+    bool isPopulation{m_OverFieldName.empty() == false};
+
+    if (m_ExcludeFrequent.empty() == false) {
+        if (m_ExcludeFrequent == ALL_TOKEN) {
+            m_ByHasExcludeFrequent = hasByField;
+            m_OverHasExcludeFrequent = isPopulation;
+        } else if (m_ExcludeFrequent == BY_TOKEN) {
+            m_ByHasExcludeFrequent = hasByField;
+        } else if (m_ExcludeFrequent == OVER_TOKEN) {
+            m_OverHasExcludeFrequent = isPopulation;
+        } else {
+            if (m_ExcludeFrequent == NONE_TOKEN) {
```

Comment:
```suggestion
            if (m_ExcludeFrequent != NONE_TOKEN) {
```

Sorry, my suggested change in the previous review was wrong - I pasted the operator from above when this one was supposed to be the reverse.

Reply:
Also, if the condition is corrected then the error message that was deleted in the second commit can be restored:

```
    LOG_ERROR(<< "Unexpected excludeFrequent value = " << m_ExcludeFrequent);
```

I know the Java should never send an invalid valid but it doesn't hurt to keep logging for the error conditions we're still detecting.